### PR TITLE
Add a lint about deriving Hash and implementing PartialEq

### LINKED ISSUE #430: Deriving only one of Hash/Eq considered harmful
If you derive Hash, but implement (Partial)Eq manually, you're probably doing it wrong, because those two need to agree for e.g. HashMap to work right.

### PR DESCRIPTION
**Pull request by mcarton**
Fix #430.

### CODE REVIEW

File: src/derive.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+use syntax::ast::{Attribute, MetaItem_};
+use utils::{match_path, span_lint_and_then};
+use utils::HASH_PATH;
+
+use rustc::middle::ty::fast_reject::simplify_type;
+
+/// **What it does:** This lint warns about deriving `Hash` but implementing `PartialEq`
+/// explicitely.
+///
+/// **Why is this bad?** The implementation of these traits must agree so it’s probably a bad idea
```
Comment (Manishearth) [inline src/derive.rs]: Could we have a more detailed explanation here? cc @Gankro

Comment (mcarton) [inline src/derive.rs]: No problem. Is it better?

File: src/derive.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+use rustc::middle::ty::fast_reject::simplify_type;
+
+/// **What it does:** This lint warns about deriving `Hash` but implementing `PartialEq`
+/// explicitely.
+///
+/// **Why is this bad?** The implementation of these traits must agree (for example for use with
+/// `HashMap`) so it’s probably a bad idea to use a default-generated `Hash` implementation  with
+/// an explicitely defined `PartialEq`. In particular, the following must hold for any type:
+/// ```rust
```
Comment (Manishearth) [inline src/derive.rs]: there should be a linebreak before this

**Comment by Manishearth** [comment]
We can also do something similar for `Copy` and `Clone` btw.

**Pull request merged**